input_reader.cpp

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <Windows.h>
+#include <tchar.h>
+#include <vector>
+#include "cpp_tstring.hpp"
+#include "console.hpp"
+#include "input_reader.hpp"
+
+std::vector<unsigned long long int> BaseInputReader::readRemainingInput() {
+	std::vector<unsigned long long int> results;
+	while (!this->inputExhausted) results.push_back(read());
+	return results;
+}
+
+FileInputReader::FileInputReader(tstring file) {
+	this->hFile = CreateFile(file.c_str(), GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+	if (GetLastError() != NO_ERROR) {
+		DWORD error = GetLastError();
+		TCHAR* arr;
+		if (!FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM, NULL, error, LANG_USER_DEFAULT, reinterpret_cast<LPTSTR>(&arr), 1, NULL)) abort();
+		console.writeError(TEXT("Failed to open the input file for reading: "));
+		console.writeError(arr);
+		console.writeError(TEXT("\n"));
+		abort();
+	}
+	unsigned long long int fileSize;
+	GetFileSizeEx(this->hFile, (PLARGE_INTEGER) & fileSize);
+	if (fileSize == 0) inputExhausted = true;
+}
+
+unsigned long long int __input_reader_util_parse_ulonglong(BaseInputReader* _this, bool(*readChar)(TCHAR*, PDWORD)) {
+	TCHAR currChar = '\0';
+	DWORD temp = 0;
+	unsigned long long int result = 0;
+	if (!readChar(&currChar, &temp)) {
+		if (GetLastError() != ERROR_HANDLE_EOF) abort();
+		temp = 0;
+	}
+	if (temp == 0) {
+		_this->inputExhausted = true;
+		throw std::exception("There's nothing more in the input.");
+		return -1;
```

Comment:
The `return` statement is redundant.

Reply:
This proves that we need an AI-powered linter

Reply:
Agreed. Finding these advanced mistakes is something that humans just aren't suitable for, that's only possible using enhanced multi-dimensional blockchain AI.